### PR TITLE
Make connector/spark tests run with spark 3.5

### PR DESCRIPTION
**Pull request by ravivj-db**
**PR Checklist**

- [x] A description of the changes is added to the description of this PR.
- [ ] If there is a related issue, make sure it is linked to this PR.
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added or modified a feature, documentation in `docs` is updated

**Description of changes**

* Removed direct client and server(Test) dependency in `connectors/spark`.
* Added a new module `serverAndClientShaded` to `build/sbt` which assembles all server and client classes (with dependencies) and shades `jackson` and `antlr` packages.
* `connectors/spark` now depends on this assembled jar for both compile and test.
* Explicitly overriding jackson and antlr dependencies in `connectors/spark` for compatibility with spark 3.5 and compatible scala version.

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -318,28 +321,90 @@ lazy val cli = (project in file("examples") / "cli")
...
     libraryDependencies ++= Seq(
-      "org.apache.spark" %% "spark-sql" % "4.0.0-preview1",
       // Test dependencies
       "org.junit.jupiter" % "junit-jupiter" % "5.10.3" % Test,
+      "org.mockito" % "mockito-core" % "5.11.0" % Test,
+      "org.mockito" % "mockito-inline" % "5.2.0" % Test,
+      "org.mockito" % "mockito-junit-jupiter" % "5.12.0" % Test,
       "net.aichler" % "jupiter-interface" % JupiterKeys.jupiterVersion.value % Test,
-      "io.delta" %% "delta-spark" % "4.0.0rc1" % Test,
+      "org.apache.spark" %% "spark-core" % sparkVersion % Test classifier "tests",
```
Comment (cloud-fan) [inline build.sbt:390]: do we need this line? We already depend on spark-sql

File: build.sbt [hunk 2]
```diff
@@ -318,28 +321,90 @@ lazy val cli = (project in file("examples") / "cli")
...
+    assembly / fullClasspath := {
+      val compileClasspath = (server / Compile / fullClasspath).value ++ (client / Compile / fullClasspath).value
+      val testClasses = (server / Test / products).value
+      compileClasspath ++ testClasses.map(Attributed.blank)
+    }
+  )
+
+val sparkVersion = "3.5.1"
 lazy val spark = (project in file("connectors/spark"))
-  .dependsOn(client % "compile->compile;test->test")
```
Comment (cloud-fan) [inline build.sbt:322]: so we don't depend on the UC client anymore? I'm wondering if it still works to package the `connectors/spark` project and use it in a spark application, as UC client is no longer a transitive dependency.

Comment (ravivj-db) [reply build.sbt:322]: Haven't tried assembling it. But client is a dependency in the sense : we assemble both server and client and that jar is both compile and test dependency for now. So wouldn't that be included in `connector/spark` assembly ?

File: build.sbt [hunk 3]
```diff
@@ -318,28 +321,95 @@ lazy val cli = (project in file("examples") / "cli")
     Test / javaOptions += s"-Duser.dir=${(ThisBuild / baseDirectory).value.getAbsolutePath}",
   )
 
+lazy val serverShaded = (project in file("serverShaded"))
```
Comment (tdas) [inline build.sbt:324]: serverAndClientShaded

File: build.sbt (hunk 1 above)
Comment (tdas) [reply build.sbt:390]: @ravivj-db good question .. does it work without this line?

File: build.sbt [hunk 4]
```diff
@@ -318,28 +321,95 @@ lazy val cli = (project in file("examples") / "cli")
...
+    libraryDependencies ++= Seq(
+      "org.apache.spark" %% "spark-sql" % sparkVersion,
+      "com.fasterxml.jackson.core" % "jackson-databind" % "2.15.0",
+      "com.fasterxml.jackson.module" %% "jackson-module-scala" % "2.15.0",
+      "com.fasterxml.jackson.core" % "jackson-annotations" % "2.15.0",
+      "com.fasterxml.jackson.core" % "jackson-core" % "2.15.0",
+      "org.antlr" % "antlr4-runtime" % "4.9.3",
+      "org.antlr" % "antlr4" % "4.9.3",
+    ).map(_.excludeAll(
+      ExclusionRule("com.fasterxml.jackson.core", "jackson-databind"),
```
Comment (tdas) [inline build.sbt:376]: i dont get what this is doing. it seems to me that this is effectively removing jackson from all the explicit jackson dependencies above.

File: build.sbt [hunk 5]
```diff
@@ -318,28 +321,95 @@ lazy val cli = (project in file("examples") / "cli")
...
+    name := s"${artifactNamePrefix}-server-shaded",
+    commonSettings,
+    skipReleaseSettings,
+    Compile / packageBin := assembly.value,
+    /*
+
+    Compile / fullClasspath := (Compile / fullClasspath).value ++ (serverModels / assembly  / fullClasspath).value,
+    Compile / fullClasspath := (Compile / fullClasspath).value ++ (client / Compile / fullClasspath).value,
+    assembly / assemblyJarName := s"${name.value}_${scalaBinaryVersion.value}-${version.value}.jar",
+     */
```
Comment (tdas) [inline build.sbt:337]: can you clean this up?

File: build.sbt [hunk 6]
```diff
@@ -318,28 +321,95 @@ lazy val cli = (project in file("examples") / "cli")
...
+    assembly / logLevel := Level.Debug,
```
Comment (tdas) [inline build.sbt:338]: warn

File: build.sbt (hunk 1 above)
Comment (ravivj-db) [reply build.sbt]: It's working without it.

File: build.sbt (hunk 3 above)
Comment (ravivj-db) [reply build.sbt]: Done

File: build.sbt (hunk 4 above)
Comment (ravivj-db) [reply build.sbt]: Removed this as not required.

File: build.sbt (hunk 5 above)
Comment (ravivj-db) [reply build.sbt]: Done

File: build.sbt (hunk 6 above)
Comment (ravivj-db) [reply build.sbt]: Done

**Review by cloud-fan — APPROVED** [review]

File: build.sbt [hunk 7]
```diff
@@ -250,6 +250,9 @@ lazy val server = (project in file("server"))
     Test / javaOptions += s"-Duser.dir=${(ThisBuild / baseDirectory).value.getAbsolutePath}"
   )
 
+
+val assemblyAndPublishLocal = taskKey[Unit]("Builds the assembly and publishes it locally")
```
Comment (tdas) [inline build.sbt:254]: why is this needed any more?

File: build.sbt [hunk 8]
```diff
@@ -320,29 +323,73 @@ lazy val cli = (project in file("examples") / "cli")
     Test / javaOptions += s"-Duser.dir=${(ThisBuild / baseDirectory).value.getAbsolutePath}",
   )
 
+lazy val serverAndClientShaded = (project in file("server-and-client-shaded"))
```
Comment (tdas) [inline build.sbt:326]: can you leave comment on what this project is used for.

Comment (tdas) [reply build.sbt:326]: also mention that this combines test classes as well.

File: project/plugins.sbt [hunk 9]
```diff
@@ -38,3 +38,7 @@ addSbtPlugin("com.etsy" % "sbt-checkstyle-plugin" % "3.1.1")
 dependencyOverrides += "com.puppycrawl.tools" % "checkstyle" % "8.43"
 
 addSbtPlugin("com.github.sbt" % "sbt-jacoco" % "3.4.0")
+
+addSbtPlugin("com.jsuereth" % "sbt-pgp" % "2.1.1")
```
Comment (tdas) [inline project/plugins.sbt:42]: what are these 2 plugins for?

File: build.sbt [hunk 10]
```diff
@@ -360,11 +407,14 @@ lazy val spark = (project in file("connectors/spark"))
...
-    },
+      case DepModuleInfo("org.glassfish", "javax.json", _) => true
+      case DepModuleInfo("org.glassfish.hk2.external", "jakarta.inject", _) => true
+      case DepModuleInfo("org.antlr", "ST4", _) => true,
+    }
   )
 
 lazy val root = (project in file("."))
-  .aggregate(serverModels, client, server, cli, spark)
+  .aggregate(serverModels, client, server, cli, spark, serverAndClientShaded)
```
Comment (tdas) [inline build.sbt:417]: is this needed? wont spark project automatically compile this project?

File: build.sbt (hunk 8 above)
Comment (ravivj-db) [reply build.sbt:326]: Sure, had added to the PR description, will add here as well.

File: build.sbt (hunk 7 above)
Comment (ravivj-db) [reply build.sbt:254]: Not needed.

File: build.sbt (hunk 10 above)
Comment (ravivj-db) [reply build.sbt:417]: I thought all modules have to be mentioned here. Why is serverModels here then ?

Comment (ravivj-db) [reply build.sbt]: Removed it, and everything's working fine.

File: build.sbt (hunk 7 above)
Comment (ravivj-db) [reply build.sbt]: Removed

File: build.sbt (hunk 8 above)
Comment (ravivj-db) [reply build.sbt:330]: Added.

File: project/plugins.sbt [hunk 11]
```diff
@@ -37,4 +37,4 @@ addSbtPlugin("com.etsy" % "sbt-checkstyle-plugin" % "3.1.1")
 // same version as Spark
 dependencyOverrides += "com.puppycrawl.tools" % "checkstyle" % "8.43"
 
-addSbtPlugin("com.github.sbt" % "sbt-jacoco" % "3.4.0")
+addSbtPlugin("com.github.sbt" % "sbt-jacoco" % "3.4.0")
```
Comment (tdas) [inline project/plugins.sbt:40]: undo this change.

Comment (ravivj-db) [reply project/plugins.sbt]: Done :).
